rust/onnxruntime-sys/build.rs

```diff
@@ -0,0 +1,420 @@
+                );
+            }
+        }
+    }
+}
+
+fn prebuilt_archive_url() -> (PathBuf, String) {
+    let triplet = Triplet {
+        os: env::var("CARGO_CFG_TARGET_OS")
+            .expect("Unable to get TARGET_OS")
+            .parse()
+            .unwrap(),
+        arch: env::var("CARGO_CFG_TARGET_ARCH")
+            .expect("Unable to get TARGET_ARCH")
+            .parse()
+            .unwrap(),
+        accelerator: env::var(ORT_RUST_ENV_GPU)
+            .unwrap_or_default()
+            .parse()
+            .unwrap(),
+    };
+
+    let prebuilt_archive = format!(
+        "onnxruntime-{}-{}.{}",
+        triplet.as_onnx_str(),
+        ORT_VERSION,
+        triplet.os.archive_extension()
+    );
+    let prebuilt_url = format!(
+        "{}/v{}/{}",
+        ORT_RELEASE_BASE_URL, ORT_VERSION, prebuilt_archive
+    );
+
+    (PathBuf::from(prebuilt_archive), prebuilt_url)
+}
+
+fn prepare_libort_dir_prebuilt() -> PathBuf {
+    let (prebuilt_archive, prebuilt_url) = prebuilt_archive_url();
+
+    let out_dir = PathBuf::from(env::var("OUT_DIR").unwrap());
+    let extract_dir = out_dir.join(ORT_PREBUILT_EXTRACT_DIR);
+    let downloaded_file = out_dir.join(&prebuilt_archive);
+
+    println!("cargo:rerun-if-changed={}", downloaded_file.display());
+
+    if !downloaded_file.exists() {
+        println!("Creating directory {:?}", out_dir);
+        fs::create_dir_all(&out_dir).unwrap();
+
+        println!(
+            "Downloading {} into {}",
+            prebuilt_url,
+            downloaded_file.display()
+        );
+        download(&prebuilt_url, &downloaded_file);
+    }
+
+    if !extract_dir.exists() {
+        println!("Extracting to {}...", extract_dir.display());
+        extract_archive(&downloaded_file, &extract_dir);
+    }
+
+    extract_dir.join(prebuilt_archive.file_stem().unwrap())
+}
+
+fn prepare_libort_dir() -> PathBuf {
+    let strategy = env::var(ORT_RUST_ENV_STRATEGY);
+    println!(
+        "strategy: {:?}",
+        strategy.as_ref().map_or_else(|_| "unknown", String::as_str)
+    );
+    match strategy.as_ref().map(String::as_str) {
+        Ok("download") => prepare_libort_dir_prebuilt(),
+        Ok("system") => PathBuf::from(match env::var(ORT_RUST_ENV_SYSTEM_LIB_LOCATION) {
+            Ok(p) => p,
+            Err(e) => {
+                panic!(
+                    "Could not get value of environment variable {:?}: {:?}",
+                    ORT_RUST_ENV_SYSTEM_LIB_LOCATION, e
+                );
+            }
+        }),
+        Ok("compile") | Err(_) => prepare_libort_dir_compiled(),
+        _ => panic!("Unknown value for {:?}", ORT_RUST_ENV_STRATEGY),
+    }
+}
+
+fn prepare_libort_dir_compiled() -> PathBuf {
+    let mut config = cmake::Config::new("../../cmake");
+
+    config.define("onnxruntime_BUILD_SHARED_LIB", "ON");
+
+    if env::var(ORT_RUST_ENV_GPU).unwrap_or_default().parse() == Ok(Accelerator::Gpu) {
+        config.define("onnxruntime_USE_CUDA", "ON");
```

Comment:
Does user need download onnxruntime_providers_* libraries as userself when enabled `onnxruntime_USE_CUDA` ?

Reply:
This part of the code will build the onnxruntime, and this flag will add cuda support, so nothing will be needed to be downloaded.